### DISCUSSION TITLE
Matter

### DISCUSSION THREAD

**Discussion by Nezaemmy** [espressif/arduino-esp32#10531] [Q&A]
is there a way to name Lights Endpoint in the code rather than to rename in set up on the device. 
always come up on the device as  (light, light1, light2, etc)
@SuGlider

**Comment by SuGlider** [comment]
> is there a way to name Lights Endpoint in the code rather than to rename in set up on the device. 
> always come up on the device as  (light, light1, light2, etc)
> @SuGlider 

This is something that I have to investigate.
If a solution is found, we shall add an API for it.

Thanks for asking!

**Reply by Nezaemmy** [reply]
Thank you

**Comment by Nezaemmy** [comment]
@SuGlider MatterOnOffLight can support more than 3 lights endpoint?
Is there a way to increase the number of them?

**Reply by SuGlider** [reply]
Yes, it is possible. There is a limit of endpoints for a single node, which, I think that may be around 16.

Using this example: https://github.com/espressif/arduino-esp32/blob/release/v3.1.x/libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino

In the sketch you can add one or more `MatterOnOffLight OnOffLight_n;` declarations for each extra light _**\<\_n is the sequencial number of that light\>**_, or just name it the way you want.
In the code it is necessary to add `OnOffLight_n.begin(bool initialState);` and also `OnOffLight_n.onChangeOnOff(setLightOnOff_n);` with the repective code for `setLightOnOff_n()`.

For testing, just follow the example code. It prints, in the serial console, which light is turned on and off.

Summary: it is just necessary to declare object of the endpoint type and execute _**obejct.begin()**_ before calling _**Matter.begin()**_ in order to add another light / any future endpoint to the same node.

**Reply by Nezaemmy** [reply]
> Yes, it is possible. There is a limit of endpoints for a single node, which, I think that may be around 16.
> 
> Using this example: https://github.com/espressif/arduino-esp32/blob/release/v3.1.x/libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino
> 
> In the sketch you can add one or more `MatterOnOffLight OnOffLight_n;` declarations for each extra light _**<_n is the sequencial number of that light>**_, or just name it the way you want. In the code it is necessary to add `OnOffLight_n.begin(bool initialState);` and also `OnOffLight_n.onChangeOnOff(setLightOnOff_n);` with the repective code for `setLightOnOff_n()`.
> 
> For testing, just follow the example code. It prints, in the serial console, which light is turned on and off.
> 
> Summary: it is just necessary to declare object of the endpoint type and execute _**obejct.begin()**_ before calling _**Matter.begin()**_ in order to add another light / any future endpoint to the same node.

You are right by default the endpoints for a single node are 16, I tested it. 
I wanted to use 20 endpoints for a single node, I found in ESP MATTER repository to change the number of endpoints by changing this CONFIG_ESP_MATTER_MAX_DYNAMIC_ENDPOINT_COUNT=16, but in arduino-esp32 I didn't found this option.
do you have plan to include this option on arduino-esp32 so that we can be able to set number of endpoints?

**Comment by Nezaemmy** [comment]
@SuGlider I'm using the matter library to test all scenarios that can happen in daily use.
  For the MatterOnOffLight example, I tested 3 scenario 
1. when the device is configured and connected to WiFi you can use UI or Physical button to control Light, which is OK.
2. When WiFi is disconnected while the Esp board is ON and configured, you can not use UI but the physical button can control Light, which is OK.
3. When WiFi is disconnected and you restart the Esp board while the board is already configured, Matter can not begin even if you comment this lines.
```
//while (WiFi.status() != WL_CONNECTED) {
  //  delay(500);
  //  Serial.print(".");
 // }
```
This kind of scenario can happen any time when the access point is damaged and restart esp board.
 It will be OK to be able to use physical button to control the light because the board was already configured.

**Reply by SuGlider** [reply]
This sketch shall also fix scenario 3:

``` cpp
// Matter Manager
#include <Matter.h>
#include <WiFi.h>
#include <Preferences.h>

// List of Matter Endpoints for this Node
// On/Off Light Endpoint
#include <MatterOnOffLight.h>
MatterOnOffLight OnOffLight;

// it will keep last OnOff state stored, using Preferences
Preferences lastStatePref;

// set your board LED pin here
#ifdef LED_BUILTIN
const uint8_t ledPin = LED_BUILTIN;
#else
const uint8_t ledPin = 2;  // Set your pin here if your board has not defined LED_BUILTIN
#warning "Do not forget to set the LED pin"
#endif

// set your board USER BUTTON pin here
const uint8_t buttonPin = 0;  // Set your pin here. Using BOOT Button. C6/C3 use GPIO9.

// WiFi is manually set and started
const char *ssid = "your-ssid";          // Change this to your WiFi SSID
const char *password = "your-password";  // Change this to your WiFi password

// Matter Protocol Endpoint Callback
bool setLightOnOff(bool state) {
  Serial.printf("User Callback :: New Light State = %s\r\n", state ? "ON" : "OFF");
  if (state) {
    digitalWrite(ledPin, HIGH);
  } else {
    digitalWrite(ledPin, LOW);
  }
  // store last OnOff state for when the Light is restarted / power goes off
  lastStatePref.putBool("lastOnOffState", state);
  // This callback must return the success state to Matter core
  return true;
}

void buttonTask(void *arg) {
  // Button control
  uint32_t button_time_stamp = 0;                 // debouncing control
  bool button_state = false;                      // false = released | true = pressed
  const uint32_t debouceTime = 250;               // button debouncing time (ms)
  const uint32_t decommissioningTimeout = 10000;  // keep the button pressed for 10s to decommission the light
  bool *lightState = (bool *)arg;
  // task loop
  Serial.println("Starting Button Task");
  while (true) {
    // A button is also used to control the light
    // Check if the button has been pressed
    if (digitalRead(buttonPin) == LOW && !button_state) {
      // deals with button debouncing
      button_time_stamp = millis();  // record the time while the button is pressed.
      button_state = true;           // pressed.
    }

    // Onboard User Button is used as a Light toggle switch or to decommission it
    uint32_t time_diff = millis() - button_time_stamp;
    if (button_state && time_diff > debouceTime && digitalRead(buttonPin) == HIGH) {
      button_state = false;  // released
      // Toggle button is released - toggle the light
      Serial.println("User button released. Toggling Light!");
      if (Matter.isWiFiConnected() && Matter.isDeviceCommissioned()) {
        OnOffLight.toggle();  // Matter Controller also can see the change
      } else  {
        // update lastOnOffState value when offline
        *lightState = !*lightState;
        setLightOnOff(*lightState);
      }
      // Factory reset is triggered if the button is pressed longer than 10 seconds
      if (time_diff > decommissioningTimeout && Matter.isDeviceCommissioned()) {
        Serial.println("Decommissioning the Light Matter Accessory. It shall be commissioned again.");
        OnOffLight.setOnOff(false);  // turn the light off
        // Check Matter Light Commissioning state
        Matter.decommission();
      }
    }
    delay(debouceTime / 5);
  }
  // it will never get here...
}

void setup() {
  // Initialize the USER BUTTON (Boot button) GPIO that will act as a toggle switch
  pinMode(buttonPin, INPUT_PULLUP);
  // Initialize the LED (light) GPIO and Matter End Point
  pinMode(ledPin, OUTPUT);

  Serial.begin(115200);
  while (!Serial) {
    delay(100);
  }

  // Initialize Matter EndPoint
  lastStatePref.begin("matterLight", false);
  bool lastOnOffState = lastStatePref.getBool("lastOnOffState", true);
  OnOffLight.begin(lastOnOffState);
  OnOffLight.onChangeOnOff(setLightOnOff);
  Serial.printf("Initial state: %s\r\n", OnOffLight.getOnOff() ? "ON" : "OFF");
  setLightOnOff(lastOnOffState);  // configure the Light based on initial state

  // start button task with a higher priority than Arduino loop Task
  xTaskCreate(buttonTask, "Button Task", 2048, &lastOnOffState, 2, NULL);

  // We start by connecting to a WiFi network
  Serial.print("Connecting to ");
  Serial.println(ssid);
  // enable IPv6
  WiFi.enableIPv6(true);
  // Manually connect to WiFi
  WiFi.begin(ssid, password);
  // Wait for connection
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  Serial.println("\r\nWiFi connected");
  Serial.println("IP address: ");
  Serial.println(WiFi.localIP());
  delay(500);

  // Matter beginning - Last step, after all EndPoints are initialized
  Matter.begin();
  // Check Matter Commissioning state
  if (!Matter.isDeviceCommissioned()) {
    // Matter APP commissioning help message
    Serial.println("");
    Serial.println("Matter Node is not commissioned yet.");
    Serial.println("Initiate the device discovery in your Matter environment.");
    Serial.println("Commission it to your Matter hub with the manual pairing code or QR code");
    Serial.printf("Manual pairing code: %s\r\n", Matter.getManualPairingCode().c_str());
    Serial.printf("QR code URL: %s\r\n", Matter.getOnboardingQRCodeUrl().c_str());
    // waits for Matter Light Commissioning.
    uint32_t timeCount = 0;
    while (!Matter.isDeviceCommissioned()) {
      delay(100);
      if ((timeCount++ % 50) == 0) {  // 50*100ms = 5 sec
        Serial.println("Matter Node not commissioned yet. Waiting for commissioning.");
      }
    }
  }
  // Set Matter Light state to the last known on-off state
  OnOffLight = lastOnOffState;
  Serial.println("Matter Node is commissioned and connected to Wi-Fi. Ready for use.");
}
void loop() {
  // The accessory may be removes/decommissioned by the Controller/APP
  if (!Matter.isDeviceCommissioned()) {
    // when no more Matter Fabrics are available start over
    ESP.restart();
  }
}
```

**Reply by Nezaemmy** [reply]
> This sketch shall also fix scenario 3:
> 
> ```c++
> // Matter Manager
> #include <Matter.h>
> #include <WiFi.h>
> #include <Preferences.h>
> 
> // List of Matter Endpoints for this Node
> // On/Off Light Endpoint
> #include <MatterOnOffLight.h>
> MatterOnOffLight OnOffLight;
> 
> // it will keep last OnOff state stored, using Preferences
> Preferences lastStatePref;
> 
> // set your board LED pin here
> #ifdef LED_BUILTIN
> const uint8_t ledPin = LED_BUILTIN;
> #else
> const uint8_t ledPin = 2;  // Set your pin here if your board has not defined LED_BUILTIN
> #warning "Do not forget to set the LED pin"
> #endif
> 
> // set your board USER BUTTON pin here
> const uint8_t buttonPin = 0;  // Set your pin here. Using BOOT Button. C6/C3 use GPIO9.
> 
> // WiFi is manually set and started
> const char *ssid = "your-ssid";          // Change this to your WiFi SSID
> const char *password = "your-password";  // Change this to your WiFi password
> 
> // Matter Protocol Endpoint Callback
> bool setLightOnOff(bool state) {
>   Serial.printf("User Callback :: New Light State = %s\r\n", state ? "ON" : "OFF");
>   if (state) {
>     digitalWrite(ledPin, HIGH);
>   } else {
>     digitalWrite(ledPin, LOW);
>   }
>   // store last OnOff state for when the Light is restarted / power goes off
>   lastStatePref.putBool("lastOnOffState", state);
>   // This callback must return the success state to Matter core
>   return true;
> }
> 
> void buttonTask(void *arg) {
>   // Button control
>   uint32_t button_time_stamp = 0;                 // debouncing control
>   bool button_state = false;                      // false = released | true = pressed
>   const uint32_t debouceTime = 250;               // button debouncing time (ms)
>   const uint32_t decommissioningTimeout = 10000;  // keep the button pressed for 10s to decommission the light
>   bool *lightState = (bool *)arg;
>   // task loop
>   Serial.println("Starting Button Task");
>   while (true) {
>     // A button is also used to control the light
>     // Check if the button has been pressed
>     if (digitalRead(buttonPin) == LOW && !button_state) {
>       // deals with button debouncing
>       button_time_stamp = millis();  // record the time while the button is pressed.
>       button_state = true;           // pressed.
>     }
> 
>     // Onboard User Button is used as a Light toggle switch or to decommission it
>     uint32_t time_diff = millis() - button_time_stamp;
>     if (button_state && time_diff > debouceTime && digitalRead(buttonPin) == HIGH) {
>       button_state = false;  // released
>       // Toggle button is released - toggle the light
>       Serial.println("User button released. Toggling Light!");
>       if (Matter.isWiFiConnected() && Matter.isDeviceCommissioned()) {
>         OnOffLight.toggle();  // Matter Controller also can see the change
>       } else  {
>         // update lastOnOffState value when offline
>         *lightState = !*lightState;
>         setLightOnOff(*lightState);
>       }
>       // Factory reset is triggered if the button is pressed longer than 10 seconds
>       if (time_diff > decommissioningTimeout && Matter.isDeviceCommissioned()) {
>         Serial.println("Decommissioning the Light Matter Accessory. It shall be commissioned again.");
>         OnOffLight.setOnOff(false);  // turn the light off
>         // Check Matter Light Commissioning state
>         Matter.decommission();
>       }
>     }
>     delay(debouceTime / 5);
>   }
>   // it will never get here...
> }
> 
> void setup() {
>   // Initialize the USER BUTTON (Boot button) GPIO that will act as a toggle switch
>   pinMode(buttonPin, INPUT_PULLUP);
>   // Initialize the LED (light) GPIO and Matter End Point
>   pinMode(ledPin, OUTPUT);
> 
>   Serial.begin(115200);
>   while (!Serial) {
>     delay(100);
>   }
> 
>   // Initialize Matter EndPoint
>   lastStatePref.begin("matterLight", false);
>   bool lastOnOffState = lastStatePref.getBool("lastOnOffState", true);
>   OnOffLight.begin(lastOnOffState);
>   OnOffLight.onChangeOnOff(setLightOnOff);
>   Serial.printf("Initial state: %s\r\n", OnOffLight.getOnOff() ? "ON" : "OFF");
>   setLightOnOff(lastOnOffState);  // configure the Light based on initial state
> 
>   // start button task with a higher priority than Arduino loop Task
>   xTaskCreate(buttonTask, "Button Task", 2048, &lastOnOffState, 2, NULL);
> 
>   // We start by connecting to a WiFi network
>   Serial.print("Connecting to ");
>   Serial.println(ssid);
>   // enable IPv6
>   WiFi.enableIPv6(true);
>   // Manually connect to WiFi
>   WiFi.begin(ssid, password);
>   // Wait for connection
>   while (WiFi.status() != WL_CONNECTED) {
>     delay(500);
>     Serial.print(".");
>   }
>   Serial.println("\r\nWiFi connected");
>   Serial.println("IP address: ");
>   Serial.println(WiFi.localIP());
>   delay(500);
> 
>   // Matter beginning - Last step, after all EndPoints are initialized
>   Matter.begin();
>   // Check Matter Commissioning state
>   if (!Matter.isDeviceCommissioned()) {
>     // Matter APP commissioning help message
>     Serial.println("");
>     Serial.println("Matter Node is not commissioned yet.");
>     Serial.println("Initiate the device discovery in your Matter environment.");
>     Serial.println("Commission it to your Matter hub with the manual pairing code or QR code");
>     Serial.printf("Manual pairing code: %s\r\n", Matter.getManualPairingCode().c_str());
>     Serial.printf("QR code URL: %s\r\n", Matter.getOnboardingQRCodeUrl().c_str());
>     // waits for Matter Light Commissioning.
>     uint32_t timeCount = 0;
>     while (!Matter.isDeviceCommissioned()) {
>       delay(100);
>       if ((timeCount++ % 50) == 0) {  // 50*100ms = 5 sec
>         Serial.println("Matter Node not commissioned yet. Waiting for commissioning.");
>       }
>     }
>   }
>   // Set Matter Light state to the last known on-off state
>   OnOffLight = lastOnOffState;
>   Serial.println("Matter Node is commissioned and connected to Wi-Fi. Ready for use.");
> }
> void loop() {
>   // The accessory may be removes/decommissioned by the Controller/APP
>   if (!Matter.isDeviceCommissioned()) {
>     // when no more Matter Fabrics are available start over
>     ESP.restart();
>   }
> }
> ```

Thank you @SuGlider
I made a project using this Matter library, it has shift registers for input and output, sensors and automation lights based on time and it is working well. the issue I have, it was that on the scenario 3.
Is it possible to change some condition so that if ``` Matter.isDeviceCommissioned() ``` even if ``` Matter.isWiFiConnected() ``` or ``` !Matter.isWiFiConnected() ``` so that MatterOnOffLight.ino example can work?
I ask this because I built the project based on MatterOnOffLight.ino example.

Thank you for your support

**Reply by SuGlider** [reply]
Do you mean, a  simple `Matter.isDeviceCommissioned()` / `Matter.isWiFiConnected()` change based on MatterOnOffLight.ino example to make it work with the scenario 3?

**Reply by Nezaemmy** [reply]
Any change in Matter src files that can make scenario 3 work without changing the structure of MatterOnOffLight.ino example.

**Reply by SuGlider** [reply]
No change in src code would make scenario 3 work. 
Matter depends on IPv6 (WiFi/Thread/ETH/PPP) connection. 
Matter is a network protocol.

**Reply by Nezaemmy** [reply]
Thank you so much for the information.
I was wondering if scenario 2 was able to work without internet, the scenario 3 can work too.

**Reply by SuGlider** [reply]
Please detail how the second scenario was tested. It is not possible to control the light using the APP when there is no WiFi working.

**Reply by SuGlider** [reply]
The board button, based on the example, would work after connecting to WiFi.

**Reply by Nezaemmy** [reply]
@SuGlider I'm using MatterOnOffLight.ino example,
step 1. with my iPhone and esp32s3 connected to the same WiFi (with iPhone running ios18 you can setup matter device without homehub)
2. I can use UI or a button to control LED (the devices are both connected in the same WiFi )
3. Turn OFF WiFi device(access point) only. don't disconnect the esp32 board, You can be able continue use the button to control the LED even if there is no Wifi.

**Comment by Nezaemmy** [comment]
For my understanding, if the matter is already commissioned that is the reason why I continue to use the button even if there is no WiFi.
What it can't work is UI because the board is not connected.

The issue comes when you restart the esp32 board when no internet. 
For my understanding, as before the device run without internet (before restarting), the device should check if it is already commissioned before and let the code run locally.

**Comment by SuGlider** [comment]
>3. Turn OFF WiFi device(access point) only. don't disconnect the esp32 board, You can be able continue use the button to control the LED even if there is no Wifi.

Yes, it has been commissioned and WiFi was active. After turning off WiFi AP, the device will continue to work using the esp32s3 onboard button because it will update its internal matter clustter information, but it won't be synchonized with the matter controller or iphone APP.

But if the WiFi AP is off and reset button is pressed, it won't connect and the sketch will be kept in the WiFi connection Loop until a suitable WiFi SSID is found and it is able to connect to the network.

In that case the sketch I presented in some other place [here](https://github.com/espressif/arduino-esp32/discussions/10531#discussioncomment-11142747) solves it. That sketch runs a separated Task just for the button that detects when Matter is not active and overrides it.

**Comment by Nezaemmy** [comment]
@SuGlider this was my question, if esp32 can able to update its internal cluster information without WiFi because it has been commissioned.
 that it is why I asked you if there  something like this  ``` if( matter.isCommisioned && !matter.isWifiConnected )``` to let the board use internal cluster. plus removing the while loop to let the system continue without internet connection.
```
  //while (WiFi.status() != WL_CONNECTED) {
  //delay(500);
  //  Serial.print(".");
 // }
```
the project I made all code run in ``` void loop() ``` to use thishttps://github.com/espressif/arduino-esp32/discussions/10531#discussioncomment-11142747  type will somehow not easy for the big project, because the main code will be in ```void setup ``` .

Thank you for the quick responce and guidance

**Reply by SuGlider** [reply]
Matter, as a protocol, will first stablish point to point connection from the Accessory to the Controller. That connection may be interrupted.

But when it has not yet stablished such connection, the cluster can't be internally manipulated.

Therefore after a reset, it will fail to call the internal attribute change callback to turn onn/off the GPIO that controls the physical light.

**Reply by Nezaemmy** [reply]
Ok, I understand the reason
